src/Compilers/CSharp/Portable/Binder/Binder_WithExpression.cs

```diff
@@ -0,0 +1,184 @@
+    {
+        private BoundExpression BindWithExpression(WithExpressionSyntax syntax, DiagnosticBag diagnostics)
+        {
+            var receiver = BindRValueWithoutTargetType(syntax.Receiver, diagnostics);
+
+            var receiverType = receiver.Type;
+            if (receiverType is null || receiverType.IsVoidType())
+            {
+                diagnostics.Add(ErrorCode.ERR_InvalidWithReceiverType, syntax.Receiver.Location);
+                return BadExpression(syntax, receiver);
+            }
+
+            if (receiverType.IsErrorType())
+            {
+                return BadExpression(syntax, receiver);
+            }
+
+            // PROTOTYPE: The receiver type must have a single declared instance method called 'With'
+            var members = receiverType.GetMembers("With");
+            MethodSymbol? withMethod = null;
+            if (members.Length == 1 && members[0] is MethodSymbol m)
+            {
+                withMethod = m;
+            }
+            else if (members.Length > 1)
+            {
+                // If there are multiple With methods, exclude overrides and
+                // look for a single remaining one
+                foreach (var member in members)
+                {
+                    if (member is MethodSymbol { IsOverride: false } method)
+                    {
+                        if (withMethod is null)
+                        {
+                            withMethod = method;
+                        }
+                        else
+                        {
+                            withMethod = null;
+                            break;
+                        }
+                    }
+                }
+            }
+
+            HashSet<DiagnosticInfo>? useSiteDiagnostics = null;
+            var lookupResult = LookupResult.GetInstance();
+            ImmutableArray<Symbol?> withMembers;
+
+            if (withMethod is null)
+            {
+                diagnostics.Add(ErrorCode.ERR_NoSingleWithMethod, syntax.Receiver.Location, receiverType);
+                withMembers = default;
+            }
+            else
+            {
+                // Check return type
+                if (!withMethod.ReturnType.IsEqualToOrDerivedFrom(
+                    receiverType,
+                    TypeCompareKind.ConsiderEverything,
+                    ref useSiteDiagnostics))
+                {
+                    diagnostics.Add(
+                        ErrorCode.ERR_ContainingTypeMustDeriveFromWithReturnType,
+                        syntax.Receiver.Location,
+                        receiverType,
+                        withMethod.ReturnType);
+                }
+
+                // Build WithMethod member list
+                var matchingMembers = ArrayBuilder<Symbol?>.GetInstance(withMethod.ParameterCount); 
+                foreach (var p in withMethod.Parameters)
+                {
+                    this.LookupMembersInType(
```

Comment:
It might be good to add an overload of this that takes an array of names, that way we don't have to do a full lookup of members on every parameter name.

Reply:
Or maybe a set of names, so that lookup time is constant.